compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3013,6 +3013,28 @@ public void visit(BLangTypeInit cIExpr) {
                 if (!cIExpr.initInvocation.argExprs.isEmpty()) {
                     BLangExpression iteratorExpr = cIExpr.initInvocation.argExprs.get(0);
                     BType constructType = checkExpr(iteratorExpr, env, symTable.noType);
+                    BType closeableIteratorType = symTable.langQueryModuleSymbol.scope
+                            .lookup(names.fromString(BLangCompilerConstants.ABSTRACT_CLOSEABLE_ITERATOR_CLASS))
+                            .symbol.type;
+                    BType iteratorType = symTable.langQueryModuleSymbol.scope
+                            .lookup(names.fromString(BLangCompilerConstants.ABSTRACT_ITERATOR_CLASS)).symbol.type;
+                    BAttachedFunction closeFunc = types.getAttachedFuncFromObject((BObjectType) constructType,
+                            BLangCompilerConstants.CLOSE_FUNC);
+                    if (closeFunc != null) {
+                        if (!types.isAssignable(constructType, closeableIteratorType)) {
+                            dlog.error(iteratorExpr.pos, DiagnosticCode.INVALID_STREAM_CONSTRUCTOR_CLOSEABLE_ITERATOR,
+                                    constructType);
```

Comment:
Can't we reuse `DiagnosticCode.INCOMPATIBLE_TYPES` for these?

Reply:
improved in: bf7f95d592214df4df84a89e0b02cee38e7cb162